plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -779,9 +789,24 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
          auto infile = std::ifstream(my->snapshot_path->generic_string(), (std::ios::in | std::ios::binary));
          auto reader = std::make_shared<istream_snapshot_reader>(infile);
          reader->validate();
-         reader->read_section<genesis_state>([this]( auto &section ){
-            section.read_row(my->chain_config->genesis);
+         chain_snapshot_header header;
+         reader->read_section<chain_snapshot_header>([&header]( auto& section ){
+            section.read_row(header);
+            header.validate();
          });
+
+         const auto version = header.version;
+
+         using v2 = legacy::snapshot_global_property_object_v2;
+         if (std::clamp(version, v2::minimum_version, v2::maximum_version) == version ) {
+            my->chain_config->genesis.emplace();
+            reader->read_section<genesis_state>([this]( auto &section ){
+               section.read_row(*my->chain_config->genesis);
+            });
+            my->chain_id = my->chain_config->genesis->compute_chain_id();
+            my->chain_id_source = "snapshot";
+         }
+
```

Comment:
With that new function created, simply call (passing `reader` into it) and store the returned `chain_id_type` into a local variable called `snapshot_chain_id`.